waku/v2/discv5/discover_test.go

```diff
@@ -102,3 +152,276 @@ func TestDiscV5(t *testing.T) {
 
 }
+
+func TestDiscV5WithCapabilitiesFilter(t *testing.T) {
+
+	// H1
+	host1, _, prvKey1 := tests.CreateHost(t)
+	udpPort1, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	ip1, _ := tests.ExtractIP(host1.Addrs()[0])
+	l1, err := tests.NewLocalnode(prvKey1, ip1, udpPort1, wenr.NewWakuEnrBitfield(true, true, true, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn1 := NewTestPeerDiscoverer()
+	d1, err := NewDiscoveryV5(prvKey1, l1, peerconn1, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort1)))
+	require.NoError(t, err)
+	d1.SetHost(host1)
+
+	// H2
+	host2, _, prvKey2 := tests.CreateHost(t)
+	ip2, _ := tests.ExtractIP(host2.Addrs()[0])
+	udpPort2, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	l2, err := tests.NewLocalnode(prvKey2, ip2, udpPort2, wenr.NewWakuEnrBitfield(true, true, false, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn2 := NewTestPeerDiscoverer()
+	d2, err := NewDiscoveryV5(prvKey2, l2, peerconn2, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort2)))
+	require.NoError(t, err)
+	d2.SetHost(host2)
+
+	// H3
+	host3, _, prvKey3 := tests.CreateHost(t)
+	ip3, _ := tests.ExtractIP(host3.Addrs()[0])
+	udpPort3, err := tests.FindFreeUDPPort(t, "127.0.0.1", 3)
+	require.NoError(t, err)
+	l3, err := tests.NewLocalnode(prvKey3, ip3, udpPort3, wenr.NewWakuEnrBitfield(true, true, false, true), nil, utils.Logger())
+	require.NoError(t, err)
+	peerconn3 := NewTestPeerDiscoverer()
+	d3, err := NewDiscoveryV5(prvKey3, l3, peerconn3, prometheus.DefaultRegisterer, utils.Logger(), WithUDPPort(uint(udpPort3)))
+	require.NoError(t, err)
+	d3.SetHost(host3)
+
+	defer d1.Stop()
+	defer d2.Stop()
+	defer d3.Stop()
+
+	err = d1.Start(context.Background())
+	require.NoError(t, err)
+
+	err = d2.Start(context.Background())
+	require.NoError(t, err)
+	// Set boot nodes for node2 after the DiscoveryV5 was created
+	err = d2.SetBootnodes([]*enode.Node{d1.localnode.Node()})
+	require.NoError(t, err)
+
+	err = d3.Start(context.Background())
+	require.NoError(t, err)
+	// Set boot nodes for node3 after the DiscoveryV5 was created
+	err = d3.SetBootnodes([]*enode.Node{d2.localnode.Node()})
+	require.NoError(t, err)
+
+	// Desired node capabilities
+	filterBitfield := wenr.NewWakuEnrBitfield(false, false, true, false)
+	iterator3, err := d3.PeerIterator(FilterCapabilities(filterBitfield))
+	require.NoError(t, err)
+	require.NotNil(t, iterator3)
+
+	time.Sleep(2 * time.Second)
+
+	// Check node were discovered by automatic discovery
+	require.True(t, peerconn3.HasPeer(host1.ID()) && peerconn3.HasPeer(host2.ID()))
+
+	peers, err := discoverFilterOnDemand(iterator3, 1)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(peers))
+
+	// Host1 has store support while host2 hasn't
+	require.Equal(t, host1.ID().String(), peers[0].AddrInfo.ID.String())
+
+	d3.Stop()
+	peerconn3.Clear()
+
+}
+
+func TestDiscV5WithShardFilter(t *testing.T) {
```

Comment:
Shard filtering is alredy done in defaultPredicate. If we just write a test to use it, then we should be good.
https://github.com/waku-org/go-waku/blob/826c7fb924e1f070e414773d9bee76998ad479aa/waku/v2/discv5/discover.go#L385
Unless i am missing something else, this test is only checking discv5 filtering with shards right.

Reply:
Again coming back to my point that we don't need to simulate some code if similar code already exists as another unit. Otherwise test code will become more than actual code and it would be very difficult to maintain.